File: DevEnvironment/CANMessageGenerator/src/main/java/canmessagegenerator/core/JitterMessageData.java
package canmessagegenerator.core;

import canmessagegenerator.interfaces.IJitterMessageData;
import java.time.Duration;
import java.time.Instant;

/**
 * The JitterMessageData class implements the jitter message data.
 */
public final class JitterMessageData implements IJitterMessageData {
    private static final String TRANSMITTING_TIME_TEXT = "TransmittingTime:";
    private static final String DEVIATION_TEXT = "Deviation:";
    private static final String NUMBER_OF_MESSAGES_SENT_TEXT = "numberOfMessagesSent:";
    private static final String NUMBER_OF_MESSAGES_RECEIVED_TEXT = "numberOfMessagesReceived:";

    private final Duration transmittingTime;
    private Instant lastSendingTime;
    private long numberOfMessagesSent;
    private long numberOfMessagesReceived;
    private double deviation;

    /**
     * Creates a new message data.
     */
    public static IJitterMessageData make(Duration transmittingTime) {
        return new JitterMessageData(transmittingTime);
    }

    /**
     * The JitterMessageData constructor.
     */
    private JitterMessageData(Duration transmittingTime) {
        this.transmittingTime = transmittingTime;
    }

    /**
     * Gets the transmitting time.
     */
    @Override
    public Duration getTransmittingTime() {
        return this.transmittingTime;
    }

    /**
     * Gets the last sending time of a message.
     */
    @Override
    public Instant getMessageLastSendingTime() {
        return this.lastSendingTime;
    }

    /**
     * Sets the last sending time of a message.
     */
    @Override
    public void setMessageLastSendingTime(Instant time) {
        this.lastSendingTime = time;
    }

    /**
     * Gets the number of messages sent.
     */
    @Override
    public long getNumberOfMessagesSent() {
        return this.numberOfMessagesSent;
    }

    /**
     * Gets the number of messages received.
     */
    @Override
    public long getNumberOfMessagesReceived() {
        return this.numberOfMessagesReceived;
    }

    /**
     * Increments the number of messages sent.
     */
    @Override
    public void incrementNumberOfMessagesSent() {
        ++this.numberOfMessagesSent;
    }

    /**
     * Increments the number of messages received.
     */
    @Override
    public void incrementNumberOfMessagesReceived() {
        ++this.numberOfMessagesReceived;
    }

    /**
     * Gets the Jitter deviation.
     */
    @Override
    public double getDeviation() {
        return this.deviation;
    }

    /**
     * Sets the Jitter deviation.
     */
    @Override
    public void setDeviation(double deviation) {
        this.deviation = deviation;
    }

    /**
     * Gets the string representation of this instance.
     */
    @Override
    public String toString() {
        return
            TRANSMITTING_TIME_TEXT + " " + this.transmittingTime +
            ", " +
            DEVIATION_TEXT + " " + this.deviation +
            ", " +
            NUMBER_OF_MESSAGES_SENT_TEXT + " " + this.numberOfMessagesSent +
            ", " +
            NUMBER_OF_MESSAGES_RECEIVED_TEXT + " " + this.numberOfMessagesReceived;
    }
}

File: DevEnvironmentCLion/CANMessageGenerator/core/JitterMessageData.cpp
#include "PreCompiled.h"
#include "JitterMessageData.h"
#include "JitterMessageHandlerException.h"

using namespace controllerareanetwork::messagegenerator;

const std::string JitterMessageData::transmittingTimeText = "TransmittingTime:";
const std::string JitterMessageData::deviationText = "Deviation:";
const std::string JitterMessageData::numberOfMessagesSentText = "numberOfMessagesSent:";
const std::string JitterMessageData::numberOfMessagesReceivedText = "numberOfMessagesReceived:";

/**
 * The JitterMessageData constructor.
 */
JitterMessageData::JitterMessageData(base::DurationPtr transmittingTime) :
    m_transmittingTime(transmittingTime),
    m_numberOfMessagesSent(0),
    m_numberOfMessagesReceived(0),
    m_deviation(0.0)
{
    if (!transmittingTime)
    {
        std::string errorMessage = "The transmitting time is not defined.";
        throw JitterMessageHandlerException(errorMessage);
    }
}

/**
 * The JitterMessageData destructor.
 */
JitterMessageData::~JitterMessageData()
{
}

/**
 * Gets the transmitting time.
 */
base::DurationPtr JitterMessageData::getTransmittingTime() const
{
    return m_transmittingTime;
}

/**
 * Gets the last sending time of a message.
 */
base::DateTimePtr JitterMessageData::getMessageLastSendingTime() const
{
    return m_lastSendingTime;
}

/**
 * Sets the last sending time of a message.
 */
void JitterMessageData::setMessageLastSendingTime(base::DateTimePtr time)
{
    if (!time)
    {
        std::string errorMessage = "The sending time of the message is not defined.";
        throw JitterMessageHandlerException(errorMessage);
    }

    m_lastSendingTime = time;
}

/**
 * Gets the number of messages sent.
 */
size_t JitterMessageData::getNumberOfMessagesSent() const
{
    return m_numberOfMessagesSent;
}

/**
 * Gets the number of messages received.
 */
size_t JitterMessageData::getNumberOfMessagesReceived() const
{
    return m_numberOfMessagesReceived;
}

/**
 * Increments the number of messages sent.
 */
void JitterMessageData::incrementNumberOfMessagesSent()
{
    ++m_numberOfMessagesSent;
}

/**
 * Increments the number of messages received.
 */
void JitterMessageData::incrementNumberOfMessagesReceived()
{
    ++m_numberOfMessagesReceived;
}

/**
 * Gets the Jitter deviation.
 */
double JitterMessageData::getDeviation() const
{
    return m_deviation;
}

/**
 * Sets the Jitter deviation.
 */
void JitterMessageData::setDeviation(double deviation)
{
    m_deviation = deviation;
}

/**
 * Gets the string representation of this instance.
 */
std::string JitterMessageData::toString() const
{
    std::stringstream stream;

    stream
        << transmittingTimeText << " " << m_transmittingTime
        << ", "
        << deviationText << " " << m_deviation
        << ", "
        << numberOfMessagesSentText << " " << m_numberOfMessagesSent
        << ", "
        << numberOfMessagesReceivedText << " " << m_numberOfMessagesReceived;

    return stream.str();
}
